be/src/vec/functions/function_timestamp.cpp

```diff
@@ -478,6 +478,149 @@ class FunctionUnixTimestamp : public IFunction {
     }
 };
 
+template <typename Impl>
+class FirstMonthDay : public IFunction {
+public:
+    static constexpr auto name = "first_month_day";
+    static FunctionPtr create() { return std::make_shared<FirstMonthDay<Impl>>(); }
+
+    String get_name() const override { return name; }
+
+    bool use_default_implementation_for_nulls() const override { return true; }
+
+    size_t get_number_of_arguments() const override { return 1; }
+
+    bool is_variadic() const override { return true; }
+
+    DataTypePtr get_return_type_impl(const ColumnsWithTypeAndName& arguments) const override {
+        return Impl::get_return_type_impl(arguments);
+    }
+
+    DataTypes get_variadic_argument_types_impl() const override {
+        return Impl::get_variadic_argument_types();
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) override {
+        return Impl::execute_impl(context, block, arguments, result, input_rows_count);
+    }
+};
+
+template <typename DateType>
+struct FirstMonthDayImpl {
+    static DataTypes get_variadic_argument_types() {
+        if constexpr (std::is_same_v<DateType, DataTypeDateTime>) {
+            return {std::make_shared<DataTypeDate>()};
+        } else if constexpr (std::is_same_v<DateType, DataTypeDateV2>) {
+            return {std::make_shared<DataTypeDateV2>()};
+        } else {
+            return {std::make_shared<DataTypeDateTimeV2>()};
+        }
+    }
+
+    // input DateTime and Date, return Date
+    // input DateTimeV2 and DateV2, return DateV2
+    static DataTypePtr get_return_type_impl(const ColumnsWithTypeAndName& arguments) {
+        if constexpr (std::is_same_v<DateType, DataTypeDateTime>) {
```

Comment:
If `DateType` is same as DataTypeDate, we also return DataTypeDate here

Reply:
return DataTypeDate when DateType is DataTypeDate